contracts/libraries/WeightedOracleLibrary.sol

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.5.0 <0.8.0;
+
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+
+/// @title Weighted Oracle library
+/// @notice Provides functions to integrate with different tier oracles of the same V3 pair
+library WeightedOracleLibrary {
+
+    struct TimeWeightedObservation {
+        int24 timeWeightedAverageTick;
+        uint128 timeWeightedHarmonicAverageLiquidity;
+    }
+
+    /// @notice Fetches time-weighted observations across different Uniswap V3 pools
+    /// @param pools Addresses of different Uniswap V3 pools that we want to observe
+    /// @param period Number of seconds in the past to start calculating the time-weighted observations
+    /// @return observations An array of obervations that have been time-weighted from (block.timestamp - period) to block.timestamp
+    function consult(address[] calldata pools, uint32 period) internal view returns (TimeWeightedObservation[] memory observations) {
+        require(period != 0, 'BP');
+
+        observations = new TimeWeightedObservation[](pools.length);
+        uint192 periodX160 = uint192(period) * type(uint160).max;
```

Comment:
```suggestion
        uint192 periodX160 = uint192(period) << 160;
```

Reply:
Actually, we are using `type(uint160).max` on purpose, to make sure that the liquidity value doesn't overflow.

We discussed it with @moodysalem [here](https://github.com/Uniswap/uniswap-v3-periphery/pull/146#discussion_r656352449)

There is also a test that fails in that case 😄